### PR TITLE
feat: qr code detector frame

### LINKED ISSUE #146: Add detector frame on QR scan layout
**UPDATED:**
Currently qr scan activity is just full screen activity. It gives no idea about what actually it is meant for.

Expected: 
- add detector frame on qr scan layout

I am working on it

### PR DESCRIPTION
**Pull request by mejariamol**
**[UPDATED]**
- adds qr code detector frame support in qr scan activity

fixes #146

screenshot_20170714-041746

### CODE REVIEW

**Comment by mejariamol** [comment]
@iamareebjamal please review.
Most of the QR scan activity stuff shifted in `onCreate` itself. Also removed interfere of the presenter from only view dependent stuff.

File: app/src/main/java/org/fossasia/openevent/app/qrscan/BarcodeTrackerFactory.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    BarcodeTrackerFactory(GraphicOverlay<BarcodeGraphic> barcodeGraphicOverlay) {
+        mGraphicOverlay = barcodeGraphicOverlay;
+    }
+
+    @Override
+    public Tracker<Barcode> create(Barcode barcode) {
+        BarcodeGraphic graphic = new BarcodeGraphic(mGraphicOverlay);
+        return new BarcodeGraphicTracker(mGraphicOverlay, graphic);
+    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/BarcodeTrackerFactory.java]: Stuff like these were the reason dagger was made. To remove the need of factories

**Comment by iamareebjamal** [comment]
@mejariamol If our implementation of MVP was perfect, we would have 0 "view dependent stuff", not even click handlers are allowed to be present in view. So this is not a good implementation. We are using dagger to make it modular. The only thing that will be needed here is to notify the graphics class about the barcode and already there is PublishSubject for it. Please improve this. Only a correct configuration is needed to make it fit into our paradigm.

File: app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java [hunk 2]
```diff
@@ -26,45 +20,4 @@
...
-
-    @Provides
-    CameraSource providesCameraSource(Context context, BarcodeDetector barcodeDetector) {
-        return new CameraSource
-            .Builder(context, barcodeDetector)
-            .setRequestedPreviewSize(640, 480)
-            .setRequestedFps(15.0f)
-            .setAutoFocusEnabled(true)
-            .build();
-    }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/module/BarcodeModule.java]: Removing this is not necessary, inject it wherever needed, and if they are not needed, then replace them with needed dependencies

File: app/src/main/java/org/fossasia/openevent/app/qrscan/BarcodeGraphicTracker.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    @Named("barcodeEmitter")
+    PublishSubject<Notification<Barcode>> barcodeEmitter;
+
+    BarcodeGraphicTracker(GraphicOverlay<BarcodeGraphic> overlay, BarcodeGraphic graphic) {
+        mOverlay = overlay;
+        mGraphic = graphic;
+
+        OrgaApplication
+            .getAppComponent(overlay.getContext())
+            .inject(this);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/BarcodeGraphicTracker.java]: This should not be needed, add all dependencies in a module and do constructor injection

File: app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java [hunk 4]
```diff
@@ -119,83 +143,22 @@ public void onRequestPermissionsResult(int requestCode, @NonNull String[] permis
...
                 } else {
-                    presenter.cameraPermissionGranted(false);
+                    Timber.e("No camera permission granted");
+                    DialogInterface.OnClickListener listener = (dialog, id) -> finish();
+
+                    AlertDialog.Builder builder = new AlertDialog.Builder(this);
+                    builder.setTitle("Open Event")
+                        .setMessage(R.string.no_camera_permission)
+                        .setPositiveButton(R.string.ok, listener)
+                        .show();
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: As discussed, please restore. I don't know how this helps in implementation of the detector or the development of the app. It makes it untestable

**Comment by mejariamol** [comment]
@iamareebjamal sure. I will do it in current style.

File: app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java [hunk 5]
```diff
@@ -255,4 +189,39 @@ public void showToggleDialog(long attendeeId) {
...
         });
     }
+
+    /**
+     * Creates and starts the camera.
+     */
+    private void createCameraSource() {
+        Context context = getApplicationContext();
+        barcodeDetector = new BarcodeDetector.Builder(context).build();
+        BarcodeTrackerFactory barcodeFactory = new BarcodeTrackerFactory(graphicOverlay);
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: @iamareebjamal Can't actually use the dagger for this as it takes `graphicOveraly` which is view inflated from the layout.

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: See this https://medium.com/@froger_mcs/inject-everything-viewholder-and-dagger-2-e1551a76a908

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: You can really inject anything

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: looking into it.

File: app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java [hunk 6]
```diff
@@ -138,29 +158,6 @@ private void buildCameraSource() {
         cameraSource = cameraSourceProvider.get();
     }
 
-    private void waitForSurface() {
```
Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: is been taken care in `CameraSourcePreview` class itself.

**Comment by mejariamol** [comment]
@iamareebjamal please review

File: app/src/main/java/org/fossasia/openevent/app/qrscan/ui/GraphicOverlay.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
+package org.fossasia.openevent.app.qrscan.ui;
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ui/GraphicOverlay.java]: Please change the package to `widget`

File: app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java [hunk 8]
```diff
@@ -78,17 +83,19 @@
...
 
         setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);
 
         ButterKnife.bind(this);
 
+        DaggerAppComponent.builder()
+            .androidModule(new AndroidModule(this))
+            .barcodeModule(new BarcodeModule(graphicOverlay))
+            .build()
+            .inject(this);
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/qrscan/ScanQRActivity.java]: I'd recommend that you create a new component for this, so it does not require null in constructor while initializing the app

**Comment by mejariamol** [comment]
@iamareebjamal please review

File: app/src/main/java/org/fossasia/openevent/app/common/di/component/BarcodeComponent.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import dagger.Component;
+
+@Singleton
+@Component(modules = {
+    DataModule.class,
+    NetworkModule.class,
+    PresenterModule.class,
+    BarcodeModule.class
+})
+public interface BarcodeComponent {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/component/BarcodeComponent.java]: It does not need all these modules. Basically, you inject 2 components in ScanQRActivity, one the common app component and other one this

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/di/component/BarcodeComponent.java]: cannot inject using more than one component at a time, I guess.

Comment (mejariamol) [inline app/src/main/java/org/fossasia/openevent/app/common/di/component/BarcodeComponent.java]: https://github.com/google/dagger/issues/230 @iamareebjamal

Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/openevent/app/common/di/component/BarcodeComponent.java]: OK, sounds good then

**Comment by mejariamol** [comment]
@iamareebjamal updated. Please review.

**Comment by mejariamol** [comment]
@Shailesh351 @rohanagarwal94 @arp95 please review

**Pull request merged**